acvm_backend_barretenberg/src/lib.rs

```diff
@@ -0,0 +1,331 @@
+            });
+            result
+        }
+
+    } else {
+        use std::cell::Cell;
+        use wasmer::{imports, Function, Instance, Memory, MemoryType, Module, Store, Value};
+
+        /// The number of bytes necessary to represent a pointer to memory inside the wasm.
+        const POINTER_BYTES: usize = 4;
+
+        /// The Barretenberg WASM gives us 1024 bytes of scratch space which we can use without
+        /// needing to allocate/free it ourselves. This can be useful for when we need to pass in several small variables
+        /// when calling functions on the wasm, however it's important to not overrun this scratch space as otherwise
+        /// the written data will begin to corrupt the stack.
+        ///
+        /// Using this scratch space isn't particularly safe if we have multiple threads interacting with the wasm however,
+        /// each thread could write to the same pointer address simultaneously.
+        const WASM_SCRATCH_BYTES: usize = 1024;
+
+        /// Embed the Barretenberg WASM file
+        #[derive(rust_embed::RustEmbed)]
+        #[folder = "$BARRETENBERG_BIN_DIR"]
+        #[include = "barretenberg.wasm"]
+        struct Wasm;
+
+        impl Barretenberg {
+            pub(crate) fn new() -> Barretenberg {
+                let (instance, memory) = instance_load();
+                Barretenberg { memory, instance }
+            }
+        }
+
+        #[derive(wasmer::WasmerEnv, Clone)]
+        struct Env {
+            memory: Memory,
+        }
+
+        /// A wrapper around the return value from a WASM call.
+        /// Notice, `Option<Value>` is used because not every call returns a value,
+        /// some calls are simply made to free a pointer or manipulate the heap.
+        #[derive(Debug)]
+        struct WASMValue(Option<Value>);
+
+        impl WASMValue {
+            fn value(self) -> Value {
+                self.0.unwrap()
+            }
+            fn into_i32(self) -> i32 {
+                i32::try_from(self.0.unwrap()).expect("expected an i32 value")
+            }
+        }
+
+        impl Barretenberg {
+            /// Transfer bytes to WASM heap
+            fn transfer_to_heap(&self, arr: &[u8], offset: usize) {
+                let memory = &self.memory;
+
+                #[cfg(feature = "js")]
+                {
+                    let view: js_sys::Uint8Array = memory.uint8view();
+                    for (byte_id, cell_id) in (offset..(offset + arr.len())).enumerate() {
+                        view.set_index(cell_id as u32, arr[byte_id])
+                    }
+                    return;
+                }
+
+                #[cfg(not(feature = "js"))]
+                {
+                    for (byte_id, cell) in memory.uint8view()[offset..(offset + arr.len())]
+                        .iter()
+                        .enumerate()
+                    {
+                        cell.set(arr[byte_id]);
+                    }
+                }
```

Comment:
For a diff PR: I want to explore if this can be written with an impl that supports JS and wasmer